second round of geochemistry additions

WilkAndy opened the pull request:
- free electron GWB format now parsed by gwb_reader
- free electron included in dbclass and dbconverter
- free electron and O2(g) added to unit/data/moose_testdb.json
- free electron added to GeochemicalDatabaseReader as just another secondary species
- GeochemicalDatabaseReader has the option that during database parsing the free electron can be rexpressed in terms of O2(aq) instead of O2(g).
- all python and C++ tested

Fixes #14976

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

rwcarlsen committed:
TheWarehouse: add support for cached, parametrized queries

To have a parametrized query cache do this:

```
class YourClass
{
    // base_query might include query conditions for a particular system, etc.
    Foo(Query& base_query) : _cache(base_query) { }

    void aFunction()
    {
      auto curr_subdomain = ...;
      auto curr_thread = ...;
      _cache.queryInto(_results, curr_thread, curr_subdomain),
    }

    std::vector<FooObjects *> _results;
    TheWarehouse::QueryCache<AttribThread, AttribSubdomain> _cache;
    ...
```

fixes #14260

lindsayad committed:
Prevent singular Jacobian (in theory)

In theory this commit would protect against a singular Jacobian
but we don't actually call the compute functions unless the slave
node projects on a master surface. So if we're in that situation
we're just gonna have a singular Jacobian

lindsayad committed:
Add tangential_tolerance mortar test

Refs #10950

acasagran committed:
Adding a series of regression tests for the NAFEMS 1D transient heat
conduction problem (T3) using different element types. Also adding a
documentation page describing the problem and the results. Issue #14838.

recuero committed:
Consolidate contact actions Refs. #14813

Contact pressure, penetration, and contact area are now a part of the general contact action.

recuero committed:
Apply clang format

recuero committed:
Remove documentation associated with obsolete contact actions.

recuero committed:
Clean up

recuero committed:
Avoid using 'parser().extractParams' to build object parameters.

lindsayad committed:
Add MortarInterface

lindsayad committed:
Build map from lower-d to higher-d subdomains for mortar

lindsayad committed:
Allow material reinit and swap from all SubProblem derived classes

lindsayad committed:
Reinit materials for mortar

Leverages new abstract virtuals in SubProblem for reiniting
and swapping (via sentinels) materials.

Closes #14821

lindsayad committed:
Simple test of materials use in mortar

zachmprince committed:
Local sensitivity calculation for polynomial chaos #14836

fdkong committed:
Upgrade PETSc submodule to the latest maint (3.12.4+)

Closes #14938

lindsayad committed:
Set required mat props for mortar constraints

Do this by looping over MaterialInterface::getMatPropDepedencies

lindsayad committed:
clang-format v9 changes in FEProblemBase

zachmprince committed:
Sobol sensitivity indices calculation for polynomial chaos #14836

zachmprince committed:
Allowing vector name as input for results of training of surrogate model #14710

recuero committed:
Address reviewer's comments

lindsayad committed:
Revert "Allow material reinit and swap from all SubProblem derived classes"

This reverts commit 95a5cb5ed5f740630eb75444fb13cde22afb3487.

crswong888 committed:
Require that CSVReader read a file no more than once with slight exception when using force_preic

(refs #14785)

lindsayad committed:
Use _fe_problem to reinit higher dimensional elements

I choose to do this for a couple of reasons:

1. If using _fe_problem then I don't have to change the API
of SwapBackSentinel.
2. This is much safer. This allows the MortarConstraints to use
material properties  coming from both displaced and undisplaced
materials. It's hard for me to imagine many scenarios where a user
would want to use an undisplaced material with a displaced constraint
or visa versa, but one example would be if using a
GenericConstantMaterial whose property  values are totally unrelated to
whether the mesh is displaced. It would be odd to require the user to
match the displaced setting of the constraints in this case. There is a
drawback here though, and that is speed. With this change we're going to
calculate variable values on both undisplaced and displaced meshes
regardless of whether they're needed. There's probably some checks we
can do in future commits that can limit computations done to
computations that  are needed. One idea I had was to do undisplaced and
displaced mortar constraints at the same time because then we could
potentially cut down on reinits by a factor of two...but this doesn't
really fit well into the current framework because we're looping over
mortar mesh elements (which will have fundamentally different
structures depending on whether the parent mesh is displaced or not)

lindsayad committed:
Set tangential tolerance in ContactAction for NormalNodalLMMechanicalContact

lindsayad committed:
Allow displaced elem-neighbor objects to reinit with undisplaced ref pts

Refs #14870

lindsayad committed:
Allow people to be stupid again

lindsayad committed:
Copy JxW neighbor after reiniting neighbor FE objects

Closes #14870

Commit:
Add touch command for rpm to fix yum issues.  Fixes #14907

recuero committed:
Address reviewer's comments II

lindsayad committed:
Use correct dimension to calculate JxWNeighbor

Need to make sure that We are retrienving a JxW corresponding to
the dimension of the face, not the volume

lindsayad committed:
Add different JxW test

Refs #14870

WilkAndy committed:
database reader overview doco.  Refs #14897

cpgr committed:
Python tools to read thermodynamic databases

Tools to read databases in GWB and EQ3/6 format and convert them
to an easy-to-parse json format.

Refs #14897

cpgr committed:
Thermodynamic database reader utility

This utility reads the JSON database of geochemical reactions
and enables easy access to species and reaction information.

Refs #14897

cpgr committed:
Unit tests of GeochemicalDatabaseReader

Refs #14897

WilkAndy committed:
database reader overview mentions redox in more detail.  Refs #14897

lindsayad committed:
Base node proc_id on owning element proc_id

This makes my test work in replicated mode in dbg mode in parallel
but not in distributed mode. I think I'm going to have to reivive
PR #14456 for that

WilkAndy committed:
python database converters now convert sorption stuff for the GWB format.  Small fixes, including the path to find included libraries.  Refs #14897

jessecarterMOOSE committed:
fix suffix for mint package, closes #14912

zachmprince committed:
Fixing bibliography and adding parameter references #14836

zachmprince committed:
Using references for vector and set inputs #14836

tophmatthews committed:
Added material_dependency_check flag #14914

tophmatthews committed:
Added validate_coupling flag

tophmatthews committed:
Update validate_coupling comment

Co-Authored-By: Daniel Schwen <daniel@schwen.de>

WilkAndy committed:
better error checking in gwb_reader.  add sorbtion stuff to dbclass.  Refs #14897

WilkAndy committed:
some const additions to make interfacing easier

WilkAndy committed:
unit test data/moose_testdb.json added

WilkAndy committed:
add initialization to avoid compiler warning/errors.

cpgr committed:
Change std::stod to MooseUtils::convert

Refs #14897

WilkAndy committed:
methods added to GeochemicalDatabaseReader that allow querying of internal datastructures.  stuff added to data/moose_testdb.json to enable more varied error checking

WilkAndy committed:
sorption stuff added to moose_testdb.json.  GeochemicalDatabaseReader now has getSurfaceSpecies method.  gwb_reader.py can now successfully parse gases that have a and b defined in their tsonopoulos terms.

Refs #14897

loganharbour committed:
New clang format

refs #14260

loganharbour committed:
Add copy constructor from another Query

refs #14260

zachmprince committed:
Small suggested changes for PC Sobol #14836

zachmprince committed:
Switching to std::size_t for potentially large indices in poly chaos #14710

zachmprince committed:
Removing legacy param stuff from poly chaos objects #14710

dschwen committed:
Explicitly cast bool to Real (#14918)

zachmprince committed:
Renaming PolyChaos vectorpostprocessors to PolynomialChaos #14710

zachmprince committed:
Adding Smolyak sparse grid #14833

zachmprince committed:
Documenting sparse grids #14833

WilkAndy committed:
more methods added to GeochemicalDatabaseReader that i am using in other code.  More tests added to units tests for these methods and others i added yesterday.  moose_testdb.json extended to allow further testing, etc

WilkAndy committed:
put sorbing materials into ordinary materials, and extend the latter so they can hold the required info.  ensure that sorption sites have a radius, even though it defaults to zero, in the MOOSE database

cpgr committed:
Method to print out species information

cpgr committed:
Unit test more methods in GeochemicalDatabaseReader

Refs #14897

cpgr committed:
Update documentation to describe database format

Refs #14897

Refs #14897

lindsayad committed:
Explicitly set `skip_find_neighbors=false`

libmesh/libmesh#2473 will hopefully add an allow_find_neighbors
API that has the same structure as allow_renumbering. However,
in for it to work existing users of `prepare_for_use` need to
explicitly state whether they want to `find_neighbors` or not

permcody committed:
Update remote settings again

refs #14767

dewenyushu committed:
Add documentation for contact, issue #14813

bwspenc committed:
Make TM kernels use eigenstrain_names instead of thermal_eigenstrain_name ref #14931

permcody committed:
Update modules/doc/content/help/inlhpcconnectivity.md

Co-Authored-By: Logan Harbour <loganhharbour@gmail.com>

WilkAndy committed:
surfaceSpeciesNames method added to GeochemicalDatabaseReader

zachmprince committed:
Adding Clenshaw-Curtis sparse grid #14833

WilkAndy committed:
stuff added to data/moose_testdb.json to enable testing of wacky cases

aeslaughter committed:
Fixup MooseDocs for python 3.6

(refs #14255)

aeslaughter committed:
Skip tests that rely on unicode on python 3.6 and linux

(refs #14255)

dschwen committed:
Cleanup extra node set generator (#14936)

jiangwen84 committed:
add vi solver option to phase field fracture
ref#14946

bwspenc committed:
Updates to thermal_eigenstrain_name param in porous_flow for consistency ref #14931

Commit:
Add libtirpc to CentOS instructions

Add libtirpc to CentOS pre-req Getting Started instructions.

Closes #14951

jbtompkins committed:
Added landing page and syntax documentation for XFEM module. Documented XFEM action and constraint. refs #10210

dschwen committed:
Fix python33 find/replace error (#14857)

dewenyushu committed:
Deprecate contact constraint, issue #14813:
  - GluedContactConstraint
  - MultiDContactConstraint
  - OneDContactConstraint
  - SparsityBasedContactConstraint

dschwen committed:
Add multiple coord test (#14936)

zachmprince committed:
Renaming quadrature to gaussQuadrature #14833

zachmprince committed:
Adding override to Legendre::clenshawQuadrature #14833

zachmprince committed:
Reorganizing polynomial chaos testing #14710

zachmprince committed:
Moving media from stochastic_tools module to large_media #14940

aeslaughter committed:
Add write function to pyhit

(refs #13942)

aeslaughter committed:
Add new line at end of rendered hit tree

(refs #13942)

aeslaughter committed:
Limit chigger tests to linux because of conda vtk bug

https://github.com/conda-forge/vtk-feedstock/issues/109

(refs #13942)

aeslaughter committed:
Limit peacock tests to linux because of conda vtk bug

https://github.com/conda-forge/vtk-feedstock/issues/109

(refs #13942)

aeslaughter committed:
Add newline at end when using hit format

(refs #14196)

aeslaughter committed:
Format input files for use in ImageFunction documentation

(refs #14196)

aeslaughter committed:
Remove extra print statement in bibtex.py

(refs #14196)

aeslaughter committed:
Add documentation for ImageFunction

(closes #14196)

aeslaughter committed:
Update large_media for new ImageFunction images

(refs #14196)

aeslaughter committed:
Update peacock tests that use matplotlib to set the size via rcParams

(closes #11734)
(refs #13942)

lindsayad committed:
Increase floors for module tests diffing with mac/conda

Refs #10950

jiangwen84 committed:
fixed failed heavy tests
ref #14946

WilkAndy committed:
free electron additions in geochemistry

- free electron GWB format now parsed by gwb_reader
- free electron included in dbclass and dbconverter
- free electron and O2(g) added to unit/data/moose_testdb.json
- free electron added to GeochemicalDatabaseReader as just another secondary species
- GeochemicalDatabaseReader has the option that during database parsing the free electron can be rexpressed in terms of O2(aq) instead of O2(g).
- all python and C++ tested

Fixes #14976

WilkAndy commented:
I'm quite happy with this, @cpgr.  Please have a look.  None of the python stuff alters the "free electron" stuff at all - it just passes the info through to our JSON format just the same as secondary species, minerals, etc.  But the Reader has the option to reexpress e- in terms of O2(aq) instead of O2(g) if a flag is set appropriately (default to True).   Then we don't have to manually change the database, it looks just like a translated version would be expected to look, but the user can control the re-expression if they desire (they will desire if they want the blasted free electron!)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/490359/) on 25ecb65 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14978/site)

This comment will be updated on new commits.

WilkAndy committed:
data/moose_geochemdb.json added - lots of lines, wow.  Fixes #14979

cpgr commented:
Hey Andy, I didn't realise that the full database hadn't been added! We do need a default database in there.

I do think that you should change the name to something more generic though - moose_thermodb.json, moose_geochemdb.json, geochemdb.json? It should definitely have a json extension though.

Also, can you hand edit the `original` bit in the database to get rid of your own path. Just make it generic like `moose/modules/geochemistry/python/original_dbs/gwb_llnl.dat`.

I'll have a look through the rest this afternoon.

cpgr commented:
Oh, just saw you added the minimal system bit - that might take longer to look at!

WilkAndy commented:
@cpgr, i'm just getting warmed up....

cpgr commented:
> @cpgr, i'm just getting warmed up....

Ha ha!

WilkAndy committed:
MinimalGeochemicalSystem and unit tests added
This builds the minimal geochemical system of basis species, equilibium species and kinetic species that is relevant for use-supplied basis components, mineral, gases, etc.  The data is stored in a computationally-efficient way using std::vector and DenseMatrix.
Also, add to moose_testdb.json to allow testing for wacky cases.
Fixes #14948

WilkAndy committed:
GeochemistrySpeciesSwapper utility added
Fixes #14949

WilkAndy committed:
GeochemicalModelRoot Userobject added: this provides a MOOSE interface to utils/MinimalGeochemicalSystem
GeochemicalModelInterrogator Userobject added: this enables querying and simple manipulations of the database
GeochemicalModelInterrogator Action added: this adds the GeochemicalModelInterrogator Userobject as well as setting up a simple Moose simulation with no timestepping, no output, etc

Lots of documentation and examples illustrating this functionality

Tests of the functionality

Fixes #14981

WilkAndy commented:
Woohoo, 30k lines - reviewers have fun!   Actually, about 25k lines is just a giant text file.

WilkAndy commented:
Hmmm, GeochemistryApp seems to be unknown - strange.

cpgr commented:
I think `GeochemistryApp.C` is missing a
```
registerKnownLabel("GeochemistryApp");
```
see https://github.com/idaholab/moose/blob/71a24f2ef37cc3d682a6fa70ec63f842c598c379/modules/porous_flow/src/base/PorousFlowApp.C#L33 for example.

Hopefully that is all it is.

I also noticed that `GeochemistryApp.C` is missing the moose header as well. Can you add that in too? The line count on this PR isn't big enough yet!

WilkAndy commented:
Thanks @cpgr.  What do you mean "I also noticed that GeochemistryApp.C is missing the moose header as well"?  I see "#include "Moose.h"" at the top of GeochemistryApp.C.

cpgr commented:
I meant the licence comment stuff

WilkAndy commented:
Thanks - how you knew about the registerKnownLabel i have no idea!    Now i get another problem:
```
../../../../combined/combined-opt -i pyrite.i



*** ERROR ***
A 'GeochemicalModelRoot' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```

WilkAndy commented:
Oh, i got that worked out - have to modify CombinedApp.C .   Thanks for your help @cpgr .   I thought it was something in the Makefiles, which can be rather obscure , so i went to bed last night rather puzzled.

WilkAndy committed:
add GeochemistryApp to base files.  Refs #14984

Also, run all Interrogator tests in OPT mode on civet, otherwise an assertion is thrown as noted in test/tests/interrogate_reactions/tests

WilkAndy commented:
Things i thought of overnight:

- add test of the Reader when reexpressFreeElectron = false
- add ability to control tolerance involved in invalid swap (checking for invertible matrix) to user via the Interrogators.  Should default at 1E-6 (the current hard-coded value in GeochemistrySpeciesSwapper)
- use this tolerance to set stoichiometric coefficients to zero if abs(stoi)<tolerance.  This will make for a nicer stoichiometric matrix that is surely physically correct - i've never seen real-life stoichiometric coefficients of magnitude less than about 1E-2, so 1E-6 is very tiny.  What do you think about this, @cpgr ?

cpgr commented on modules/geochemistry/include/userobjects/GeochemicalModelRoot.h:
```diff
@@ -0,0 +1,40 @@
+
+  GeochemicalModelRoot(const InputParameters & parameters);
+
+  virtual void initialize() override;
```
could mark these final instead of override so that they are never used in any derived class in the future

cpgr commented on modules/geochemistry/test/tests/interrogate_reactions/tests:
```diff
@@ -0,0 +1,112 @@
+Pyrite = - 3.5\*H2O \+ 1\*Fe\(OH\)3\(ppd\) \+ 2\*SO4-- \+ 4\*H\+ - 3.75\*O2\(aq\)  .  log10\(K\) = 221
+Pyrite = - 1\*H2O \+ 1\*Fe\+\+ \+ 2\*SO4-- \+ 2\*H\+ - 3.5\*O2\(aq\)  .  log10\(K\) = 217.4
+Pyrite = - 1\*H2O \+ 1\*Fe\+\+ \+ 2\*H2S\(aq\) - 2\*H\+ \+ 0.5\*O2\(aq\)  .  log10\(K\) = -45.39
+Pyrite = - 1\*H2O \+ 1\*Fe\+\+ \+ 1.75\*H2S\(aq\) - 1.5\*H\+ \+ 0.25\*SO4--  .  log10\(K\) = -12.54'
```
I think you missed a backslash here 😆 (just joking!)

WilkAndy replied:
Thank goodness for text editors that can do find-and-replace!

cpgr commented on modules/geochemistry/test/tests/interrogate_reactions/tests:
```diff
@@ -0,0 +1,112 @@
+    input = 'gypsum.i'
+    allow_warnings = true
+    method = 'OPT'
+    threading = '!pthreads'
```
Why are these only opt and !pthreads? Is the pthreads the same issue as the fluid properties user objects?

WilkAndy replied:
Only run in opt mode, otherwise the Outputs/perf_graph=true in the CIVET debug tests triggers:
  Assertion `common_actions.size() == 1' failed
  Should be only one common_output Action

Not sure about pthreads.  I can check....

cpgr reviewed (commented):
Wow - lots of code Andy! 

Apart from the inline comments above, I had the following qs:

What about renaming the data folders to database?

I don't get the GeochemicalModelRoot name? GeochemicalModelDefinition? 

Also the minimal in MinimalGeochemicalSystem makes me think it is the smallest possible system you could imagine, rather than all of the data for the given model. How about ModelGeochemicalSystem to go with ModelGeochemicalDatabase?

The amount of testing and documentation you have created is amazing! I think you should work from home all the time! It's really cool how fast this is coming along.

cpgr commented on modules/geochemistry/src/actions/AddGeochemicalModelInterrogatorAction.C:
```diff
@@ -0,0 +1,190 @@
+registerMooseAction(
+    "GeochemistryApp",
+    AddGeochemicalModelInterrogatorAction,
+    "add_distribution"); // During add_distribution, the GeochemicalModelInterroagor UserObject is
```
I did something similar in PorousFlowApp.C to make a new `add_joiner` task. I think something like 

```
  registerTask("add_geocheminterrogator", /*is_required=*/false);
  addTaskDependency("add_geocheminterrogator", "add_userobject");
```

in `GeochemistryApp.C` should work (with the corresponding change of swapping add_distribution with add_geocheminterrogator in  AddGeochemicalModelInterrogatorAction)

WilkAndy replied:
Nope, i just spent 45mins trying to get this to work and failed.  It's not helping that i can't find doxygen or other documentation on this stuff.  I'm not going to change anything here at present, but feel free to suggest other things to try!

cpgr replied:
Sorry! Leave it as is then - at least it works

WilkAndy replied:
Cool, sounds the easy way out!

WilkAndy commented:
Thanks for the review @cpgr .   I'll address these today.

cpgr commented:
Is the simple_diffusion test still in the tests folder after this? Want to trash it if it is?

WilkAndy commented:
> Is the simple_diffusion test still in the tests folder after this? Want to trash it if it is?

yep and yep

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/activity_ratios.md:
```diff
@@ -14,7 +14,60 @@ has an equilibrium constant $\log_{10}K = 3.42$ at 25$^{\circ}$C.  Assuming a wa
 \end{equation}
-MOOSE produces the result ????
+
+To perform this simulation using the `geochemistry` module, a [GeochemicalModelRoot](GeochemicalModelRoot.md) object just be created with the desired mineral species:
```
done

WilkAndy commented:
@cpgr, i changed "data" directories to "database".  And i changed "Root" to "Definition".  But i'm unsure of the "MinimalGeochemicalSystem" to "ModelGeochemicalSystem" change.  I understand your point, but i don't like the word Model, as it can imply "perfect" instead of "a numerical model".   Any other ideas?

WilkAndy committed:
most of @cpgr change-requests implemented - only thing not done is the Action task dependency stuff

Along with:
tests of Reader when reexpress_free_electron=false
ability for user to control tolerance in invalid swap
this tolerance is also used to set small stoichiometric coefficients to zero

Refs #14981

MinimalGeochem changed to PertinentGeochem

WilkAndy commented:
@cpgr - how about PertinentGeochemicalSystem instead of MinimalGeochemicalSystem ?   I'll go for that at present until you suggest something else.

cpgr commented:
> @cpgr - how about PertinentGeochemicalSystem instead of MinimalGeochemicalSystem ? I'll go for that at present until you suggest something else.

I'm happy with whatever you choose - was just suggesting that minimal might not be the best word.

cpgr commented:
Looks great Andy - still can't believe how much you have already done!

WilkAndy commented:
I believe this PR does not cause the MacTest failure

WilkAndy commented:
@loganharbour , this is ready for a review.  Are you  keen?

WilkAndy commented:
@loganharbour , bugging you again, heheheheheheheheheheh

loganharbour commented:
I'm slowly looking through this - the physics are new to me so most of it is learning and reading. 

While I'm still looking through it:
Why is it that the `GeochemicalModelInterrogator` only has on screen output?

WilkAndy commented:
> I'm slowly looking through this - the physics are new to me so most of it is learning and reading.
> 
> While I'm still looking through it:
> Why is it that the `GeochemicalModelInterrogator` only has on screen output?

The idea behind the `Interrogator` is that it allows geochemists to rapidly assess quite simple things.  All the calculations could quite easily be done by hand, but it's quite annoying and error-prone - nicer to use a computer.  I'm not a geochemist, but after some reading around, i think they do these sorts of simple computations frequently, to gain understanding about a system.  No mesh or timestepping are required, so on-screen output is appropriate.

WilkAndy commented:
Hi @loganharbour Just bugging you about this again....

loganharbour commented:
Wow. I'm not sure how I missed your original response to me about output. I'm sorry about that. I didn't plan on finishing up until I heard a response to that question.

I suppose I would have expected some very simple postprocessors, but now that I've looked at the output a little bit more I do see some of your point. I'll look more into this tomorrow.

WilkAndy commented:
> Wow. I'm not sure how I missed your original response to me about output. I'm sorry about that. I didn't plan on finishing up until I heard a response to that question.
> 
> I suppose I would have expected some very simple postprocessors, but now that I've looked at the output a little bit more I do see some of your point. I'll look more into this tomorrow.

There will be quite a bit more of this type of code.  Geochemists often solve non-spatial, non-temporal models, which doesn't really fit into a usual finite-element paradigm.  As i proceed with `geochemistry`, it may become clear that some of this code can be made into more standard MOOSE code, but at present i simply can't predict what the final thing is going to look like.

My next PR is getting larger every day, so please don't take too long with this one.

loganharbour commented:
> My next PR is getting larger every day, so please don't take too long with this one.

I understand 👍

WilkAndy commented:
@loganharbour , could you please advise on the following.

1. `PertinentGeochemicalSystem` defines a "minimal model" that only includes the stuff desired by the user.  It has a method `modelGeochemicalDatabaseCopy()` that is supposed to provide a *copy* of the ModelGeochemicalDatabase.  This is to ensure that no other classes actually modify the `PertinentGeochemicalSystem`'s ModelGeochemicalDatabase - they can get their own copies and do whatever they like, but the `PertinentGeochemicalSystem`'s version remains untouched.  This is important, as when i get to coding spatially-dependent stuff, each node will have its *own copy* of the ModelGeocehmicalDatabase.  Could you please confirm I've done this correctly?

2. `GeochemicalModelDefinition` is a userobject that defines the Pertinent system for the rest of MOOSE.  It provides a `getDatabase` method that returns a copy of the ModelGeochemicalSystem.  Could you please confirm that i'm not creating an additional copy here: one from the `modelGeochemicalDatabaseCopy()` and one from the `return _model.modelGeochemicalDatabaseCopy();` ?

3. `GeochemicalModelInterrogator` is a userobject that interrogates the ModelGeochemicalDatabase in various ways.  Its constructor has the line `_mgd(getUserObject<GeochemicalModelDefinition>("model_definition").getDatabase())`.  Can you confirm that i'm not creating additional copies here?  I just want a copy of the `PertinentGeochemicalSystem`'s ModelGeochemicalDatabase , not a copy of a copy, or a copy of a copy of a copy.

Hopefully these are simple questions for you.  I've been coding in other languages for a few months now, and getting back into C++ i've found i've forgotten lots of fundamental stuff.

loganharbour commented:
Good questions! This "copy" stuff was something that actually confused me a little bit when reviewing.

>     1. `PertinentGeochemicalSystem` defines a "minimal model" that only includes the stuff desired by the user.  It has a method `modelGeochemicalDatabaseCopy()` that is supposed to provide a _copy_ of the ModelGeochemicalDatabase.  This is to ensure that no other classes actually modify the `PertinentGeochemicalSystem`'s ModelGeochemicalDatabase - they can get their own copies and do whatever they like, but the `PertinentGeochemicalSystem`'s version remains untouched.  This is important, as when i get to coding spatially-dependent stuff, each node will have its _own copy_ of the ModelGeocehmicalDatabase.  Could you please confirm I've done this correctly?

Let's look at your `ModelGeochemicalDatabase` struct. All it has is a default constructor. Because it has no copy constructor, the compiler will create one for you that is a member-wise copy between objects. That is... in doing something like
```
ModelGeochemicalDatabase new_db = _model.modelGeochemicalDatabaseCopy()
```
will construct a new object inside the scope of the function call. The compiler is then smart enough to see that `new_db` is being constructed by said function and will "move" the object created in the constructor to `new_db`.

However... it is much smarter to just return const references to these database objects. With const references, you're giving the user access to the memory without the ability to do anything to it. _And_ they can use it in the same sense in the code block above to just copy the database. Therefore, you should definitely do something  like this in `PertinentGeochemicalSystem` (remove stuff in the source file, change header to this):

 ```
/// Const reference to the ModelGeochemicalDatabase, which can be used to copy it
const ModelGeochemicalDatabase & getDatabase() const { return _model; }
```

Note that I also suggest changing the method name to `getDatabase()` so that it's consistent with what's done in the UO in the bullet point below.

>     2. `GeochemicalModelDefinition` is a userobject that defines the Pertinent system for the rest of MOOSE.  It provides a `getDatabase` method that returns a copy of the ModelGeochemicalSystem.  Could you please confirm that i'm not creating an additional copy here: one from the `modelGeochemicalDatabaseCopy()` and one from the `return _model.modelGeochemicalDatabaseCopy();` ?

I'm not 100% sure on how the memory process works, that is, if the compiler is smart enough to keep moving the object created initially in `_model.modelGeochemicalDatabaseCopy()` to the other functions calling it. But regardless, I you should definitely do this with const references as I did in the first bullet to guarantee no copies. I would change this call the `GeochemicalModelDefinition::getDatabase()` to (remove stuff in the source file, change header to this):

```
/// Const reference to the ModelGeochemicalDatabase, which can be used to copy it
const ModelGeochemicalDatabase & getDatabase() const { return _model.getDatabase(); }
```

>     3. `GeochemicalModelInterrogator` is a userobject that interrogates the ModelGeochemicalDatabase in various ways.  Its constructor has the line `_mgd(getUserObject<GeochemicalModelDefinition>("model_definition").getDatabase())`.  Can you confirm that i'm not creating additional copies here?  I just want a copy of the `PertinentGeochemicalSystem`'s ModelGeochemicalDatabase , not a copy of a copy, or a copy of a copy of a copy.

Same thing here about me not being 100% sure how the memory process works (how smart the compiler is). If you make the changes I suggested in the first and second bullets, this call will still work and it will do the single copy you want.

loganharbour commented on modules/geochemistry/test/tests/interrogate_reactions/tests:
```diff
@@ -2,6 +2,11 @@
   [./invalid_swap]
     type = RunException
     input = invalid_swap.i
+    # only run in opt mode, otherwise the Outputs/perf_graph=true in the CIVET debug tests triggers:
```
This is something that we _really_ shouldn't be doing.

WilkAndy replied:
I agree we shouldn't be doing this.  Do you know how to get around CIVET putting that Outputs/perf_graph=true in these tests?

WilkAndy commented:
Thanks for the const-ref stuff @loganharbour .  I might wait until you've reviewed the whole PR before i implement the changes.

loganharbour commented:
I think we need to find some time to talk for a few minutes. I've finished my review, just missing a statement about it

We've got two outstanding issues. First, the copy stuff that we discussed previously (which you just referenced). Second, the output. As your interrogator isn't using PPs or VPPs for any sort of output, but is instead just printing to screen... this should really be handled as an Output object. This will also take care of the print out hackery that you tried to take care of with the action system and adding the console.

WilkAndy commented:
OK about Output.  I just copied the FluidPropertiesInterrogator almost verbatim and am happy to change to something better.  Can you describe in more detail what you're thinking of?   Or perhaps we can skype or something?

loganharbour commented:
> OK about Output. I just copied the FluidPropertiesInterrogator almost verbatim and am happy to change to something better. Can you describe in more detail what you're thinking of? Or perhaps we can skype or something?

Any time you can find me on Slack is fair game! I'm still a student so I'll definitely be around on the weekend as well.

WilkAndy committed:
use const references instead of copies.  Refs #14981

WilkAndy committed:
GeochemicalModelInterrogator is now an Output instead of a UserObject.

tests now run in dbg

Fixes #14981

WilkAndy commented:
Hi @loganharbour , I'm going to close this PR in favour of #15098 .   I got git all messed up with different branches and my new material that's not yet a PR, and the "next" branch, all mixed up together, and it probably wasn't helping that I was watching a movie at the same time. hmmmm.  So, I wanted to squash everything anyway, so look at #15098.  It has all the `Output` stuff implemented, as well as the const-ref stuff you mention above.

It'd be great to get that material merged because I'm finding it more and more difficult to work on my next round of stuff.
